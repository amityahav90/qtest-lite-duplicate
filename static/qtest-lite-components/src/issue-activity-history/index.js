import React from 'react';
import ReactDOM from 'react-dom';
import { IssueActivityHistory } from './issue-activity-history.component';
import '@atlaskit/css-reset';
import { IntlProvider } from 'react-intl-next';

ReactDOM.render(
	<React.StrictMode>
		<IntlProvider locale="en">
			<IssueActivityHistory />
		</IntlProvider>
	</React.StrictMode>,
	document.getElementById('root')
);
