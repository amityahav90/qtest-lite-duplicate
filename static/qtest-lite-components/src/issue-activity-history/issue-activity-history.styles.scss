.issue-activity-history-spinner {
	display: flex;
	justify-content: center;
	margin: 10px 0;
}

.issue-activity-history-container {
	.history-item {
		display: flex;
		align-items: center;

		.history-item-ml5 {
			margin-left: 5px;
		}

		& + .history-item {
			margin-bottom: 10px;
		}
	}

	.history-item-changes {
		margin-left: 50px;

		.history-item-change {
			& + .history-item-change {
				margin-top: 5px;
			}
		}
	}
}
