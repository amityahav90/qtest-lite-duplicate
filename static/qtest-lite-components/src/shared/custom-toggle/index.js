export * from './custom-toggle.component';
