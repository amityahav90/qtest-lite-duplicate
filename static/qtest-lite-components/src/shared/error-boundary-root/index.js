export * from './error-boundary-root.component';
export * from './with-error-boundary-root';
