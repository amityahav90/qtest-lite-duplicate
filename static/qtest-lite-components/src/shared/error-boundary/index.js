export * from './error-boundary.component';
export * from './with-error-boundary';
