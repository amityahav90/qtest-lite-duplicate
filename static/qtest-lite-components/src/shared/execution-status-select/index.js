export * from './execution-status-select.component';
export * from './get-execution-status-label.function';
export * from './get-execution-status-colors.function';
