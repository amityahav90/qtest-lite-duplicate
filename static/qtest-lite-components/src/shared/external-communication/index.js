export * from './external-communication-service';
