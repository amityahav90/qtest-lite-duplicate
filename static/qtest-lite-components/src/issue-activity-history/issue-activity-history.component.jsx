import React, { Fragment, useEffect, useState } from 'react';
import { useLogHook, withLogPath } from '../shared/log-service';
import { ExternalCommunicationService } from '../shared/external-communication';
import Avatar from '@atlaskit/avatar';
import moment from 'moment';
import Spinner from '@atlaskit/spinner';
import './issue-activity-history.styles.scss';
import { withErrorBoundaryRoot } from '../shared/error-boundary-root';
import DynamicTable from '@atlaskit/dynamic-table';
import i18n from '../shared/localization/i18n';
import { AppSettingsService } from '../../../../shared/app-settings-service';
import { RichTextboxView } from '../shared/rich-textbox-view';
import { camelToSentenceCase } from '../../../../shared/tools';

const externalCommunicationService = new ExternalCommunicationService();

function IssueActivityHistoryInternal() {
	const log = useLogHook();
	const [history, setHistory] = useState(undefined);
	const [error, setError] = useState(false);
	const head = {
		cells: [
			{
				key: 'step-description',
				content: 'Step Description'
			},
			{
				key: 'changed',
				content: 'Changed'
			},
			{
				key: 'original-value',
				content: 'Original Value'
			},
			{
				key: 'new-value',
				content: 'New Value'
			}
		]
	};
	const dateFormat = new AppSettingsService().getDateFormat();

	function getRelevantView(key, value) {
		if (key === 'attachments') {
			return value && value.map((item) => item.filename).join(', ');
		} else {
			return <RichTextboxView defaultValue={value.toString()} />;
		}
	}

	function readHistoryItemChangeColumn(key, stepDescription = '', changed, originalValue = '', newValue = '') {
		return {
			key,
			cells: [
				{ key: 'step-description', content: getRelevantView('', stepDescription) },
				{ key: 'changed', content: camelToSentenceCase(changed) },
				{ key: 'original-value', content: getRelevantView(changed, originalValue) },
				{ key: 'new-value', content: getRelevantView(changed, newValue) }
			]
		};
	}

	function readHistoryItemChange(change, index) {
		const changes = [];

		if (change.oldValue === undefined) {
			changes.push(
				readHistoryItemChangeColumn(`row-${index}-created`, change.newValue.summary || change.newValue.description, 'Created')
			);
		} else if (change.newValue === undefined) {
			changes.push(
				readHistoryItemChangeColumn(`row-${index}-deleted`, change.oldValue.summary || change.oldValue.description, 'Deleted')
			);
		} else {
			Object.keys(change.newValue).forEach((newValueKey) => {
				if (JSON.stringify(change.newValue[newValueKey]) !== JSON.stringify(change.oldValue[newValueKey])) {
					changes.push(
						readHistoryItemChangeColumn(
							`row-${index}-changed-${newValueKey.toLowerCase()}`,
							change.oldValue.description,
							newValueKey,
							change.oldValue[newValueKey],
							change.newValue[newValueKey]
						)
					);
				}
			});
		}

		return changes;
	}

	function readHistory(history) {
		return history.map((historyItem, index) => ({
			user: historyItem.user,
			type: historyItem.type,
			operation: historyItem.operation,
			date: historyItem.date,
			changes: historyItem.changes
				.map((change, changeIndex) => readHistoryItemChange(change, changeIndex))
				.reduce((arr, item) => {
					arr.push(...item);

					return arr;
				}, [])
		}));
	}

	useEffect(() => {
		externalCommunicationService
			.getIssueHistory()
			.then((history) => {
				setHistory(readHistory(history));
			})
			.catch((e) => {
				setError(true);
				log.error('Failed to retrieve Issue History', e);
			});
	}, []);

	if (error) {
		return <div>{i18n.t('ISSUE_ACTIVITY_HISTORY_OPERATION_FAILED')}</div>;
	}

	if (!history) {
		return (
			<div className="issue-activity-history-spinner">
				<Spinner size="xlarge" />
			</div>
		);
	}

	if (history.length === 0) {
		return i18n.t('ISSUE_ACTIVITY_HISTORY_NO_CHANGES');
	}

	return (
		<div className="issue-activity-history-container">
			{history.map((historyItem, index) => (
				<Fragment key={index}>
					<div className="history-item">
						<Avatar src={historyItem.user.avatarUrl} />
						<span className="history-item-ml5">
							<b>{historyItem.user.name}</b>
						</span>
						<span className="history-item-ml5">made changes on</span>
						<span className="history-item-ml5">{moment(historyItem.date).format(dateFormat)}</span>
					</div>
					<div className="history-item-changes">
						<DynamicTable head={head} rows={historyItem.changes} isFixedSize={true} />
					</div>
				</Fragment>
			))}
		</div>
	);
}

export const IssueActivityHistory = withLogPath(withErrorBoundaryRoot(IssueActivityHistoryInternal), 'IssueActivityHistory');
