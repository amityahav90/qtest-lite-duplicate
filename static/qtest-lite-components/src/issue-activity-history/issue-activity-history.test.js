import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { IssueActivityHistory } from './issue-activity-history.component';
import { ExternalCommunicationService } from '../shared/external-communication';

describe('Test IssueActivityHistory component', () => {
	beforeAll(() => {
		jest.mock('@forge/bridge');
	});

	test('Component rendered correctly', async () => {
		let component;

		await act(async () => {
			component = renderer.create(<IssueActivityHistory />);
		});

		expect(component.toJSON()).toMatchSnapshot();
	});

	test('Component rendered no changes message correctly', async () => {
		let component;

		const spy = jest.spyOn(ExternalCommunicationService.prototype, 'getIssueHistory').mockImplementation(() => Promise.resolve([]));

		await act(async () => {
			component = renderer.create(<IssueActivityHistory />);
		});

		expect(component.toJSON()).toMatchSnapshot();
		spy.mockRestore();
	});

	test('Component rendered error message correctly', async () => {
		let component;

		const spy = jest.spyOn(ExternalCommunicationService.prototype, 'getIssueHistory').mockImplementation(() => Promise.reject());

		await act(async () => {
			component = renderer.create(<IssueActivityHistory />);
		});

		expect(component.toJSON()).toMatchSnapshot();
		spy.mockRestore();
	});
});
