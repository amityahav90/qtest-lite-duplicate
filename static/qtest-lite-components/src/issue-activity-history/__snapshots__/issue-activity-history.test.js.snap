// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test IssueActivityHistory component Component rendered correctly 1`] = `
<div
  className="issue-activity-history-container"
>
  <div
    className="history-item"
  >
    <div
      style={
        Object {
          "display": "inline-block",
          "outline": 0,
          "position": "relative",
          "zIndex": undefined,
        }
      }
    >
      <span
        className="css-tx7wnu"
      >
        <span
          className="css-1l72gtu-AvatarImage"
        >
          <span
            aria-hidden={true}
            className="css-60ak9x-Icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" role=\\"presentation\\"><g fill=\\"currentColor\\" fill-rule=\\"evenodd\\"><path d=\\"M6 14c0-1.105.902-2 2.009-2h7.982c1.11 0 2.009.894 2.009 2.006v4.44c0 3.405-12 3.405-12 0V14z\\"/><circle cx=\\"12\\" cy=\\"7\\" r=\\"4\\"/></g></svg>",
              }
            }
            role="presentation"
            style={
              Object {
                "--icon-primary-color": "var(--ds-text-onBold, #FFFFFF)",
                "--icon-secondary-color": "var(--ds-text-lowEmphasis, #8993A4)",
              }
            }
          />
        </span>
      </span>
    </div>
    <span
      className="history-item-ml5"
    >
      <b>
        John Smith
      </b>
    </span>
    <span
      className="history-item-ml5"
    >
      made changes on
    </span>
    <span
      className="history-item-ml5"
    >
      November 29, 2021, 9:33:30 AM
    </span>
  </div>
  <div
    className="history-item-changes"
  >
    <div
      className="css-8jk8id-Container"
    >
      <table
        className="css-uyynlu"
        style={
          Object {
            "--local-dynamic-table-highlighted-bg": "var(--ds-background-selected-resting, #DEEBFF)",
            "--local-dynamic-table-hover-bg": "var(--ds-background-subtleBorderedNeutral-resting, #FAFBFC)",
            "--local-dynamic-table-hover-highlighted-bg": "var(--ds-background-selected-hover, #B3D4FF)",
            "--local-dynamic-table-row-focus-outline": "var(--ds-border-focus, #4C9AFF)",
          }
        }
      >
        <thead
          className="css-1qnlpdw-Head"
        >
          <tr>
            <th
              className="css-1dtsdxj"
              onKeyDown={[Function]}
              style={
                Object {
                  "--local-dynamic-table-text-color": "var(--ds-text-lowEmphasis, #5E6C84)",
                }
              }
            >
              <span>
                Step Description
              </span>
            </th>
            <th
              className="css-1dtsdxj"
              onKeyDown={[Function]}
              style={
                Object {
                  "--local-dynamic-table-text-color": "var(--ds-text-lowEmphasis, #5E6C84)",
                }
              }
            >
              <span>
                Changed
              </span>
            </th>
            <th
              className="css-1dtsdxj"
              onKeyDown={[Function]}
              style={
                Object {
                  "--local-dynamic-table-text-color": "var(--ds-text-lowEmphasis, #5E6C84)",
                }
              }
            >
              <span>
                Original Value
              </span>
            </th>
            <th
              className="css-1dtsdxj"
              onKeyDown={[Function]}
              style={
                Object {
                  "--local-dynamic-table-text-color": "var(--ds-text-lowEmphasis, #5E6C84)",
                }
              }
            >
              <span>
                New Value
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="css-v3yuj6"
          >
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              Index
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr
            className="css-v3yuj6"
          >
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              Index
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr
            className="css-v3yuj6"
          >
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              Index
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
            <td
              className="css-12vmjho-TableBodyCell"
              style={
                Object {
                  "--local-dynamic-table-width": undefined,
                }
              }
            >
              <div
                className="rich-textbox-view-component"
              >
                <div
                  className="sc-fQkuQJ iJqztD sc-bdVaJa jGgciL"
                >
                  <div />
                  <div
                    className="sc-bNQFlB gNbClm"
                  >
                    <div
                      className="sc-dRCTWM cMZerd"
                    >
                      <div
                        className="sc-cgHJcJ cDGnAG sc-crNyjn ktaVza"
                      >
                        <div
                          aria-label="Main content area"
                          className=""
                          role="textbox"
                        />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`Test IssueActivityHistory component Component rendered error message correctly 1`] = `
<div>
  Operation failed, please try again. If the problem persists, please contact support.
</div>
`;

exports[`Test IssueActivityHistory component Component rendered no changes message correctly 1`] = `"There are no changes here yet"`;
